### PR TITLE
ftux: birth

### PR DESCRIPTION
**Pull request by thelostone-mc**
##### Description

Build user onboarding flow 

```
- Step1  // Github login - Done
- Step2 // Metamask -  Donr
- Step3 // Get skills - Done
- Final Step // Done
 ```

##### Checklist

- [x] linter status: 100% pass
- [x] changes don't break existing behavior
- [x] commit message follows [commit guidelines](https://github.com/gitcoinco/web/blob/master/docs/CONTRIBUTING.md#step-4-commit)

##### Affected core subsystem(s)
- css
- html
- js

##### Refers/Fixes
https://github.com/gitcoinco/web/issues/441

### CODE REVIEW

**Comment by thelostone-mc** [comment]
Status : Pending sync up with @mbeacom  for help ^_^

File: app/assets/v2/images/metamask.svg [hunk 1]
```diff
@@ -0,0 +1,128 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (mbeacom) [inline app/assets/v2/images/metamask.svg]: We'll want to run `svgo` over this before we merge.  If you have `svgo` installed, you can run something like:  `svgo app/assets/v2/images/metamask.svg` or simply run `make compress-images`

File: app/dashboard/templates/onboard.html [hunk 2]
```diff
@@ -0,0 +1,175 @@
+{% comment %}
+    Copyright (C) 2017 Gitcoin Core
```
Comment (mbeacom) [inline app/dashboard/templates/onboard.html]: Can you update the copyright to `2018` and bring the indent in by one.

File: app/dashboard/templates/onboard.html [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+                    <h3>
+                      <i class="far fa-check-circle"></i>
+                      {% trans "You've logged in!" %}
+                    </h3>
+                    <div class="font-caption row">
+                      <div class="col-12">
+                        <img class="github-avatar" src="{% url "avatar" %}?repo=https://github.com/{{ github_handle }}">
+                        <span>
+                          <p>{{ github_handle }}</p>
+                          <a href="{% url "github:github_logout" %}?redirect_uri={{ request.get_full_path }}">
```
Comment (mbeacom) [inline app/dashboard/templates/onboard.html]: This `url` view reference will need updated with the new views specified in `nav_auth.html`

File: app/dashboard/templates/onboard.html [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+                            {% trans "Logout" %}
+                          </a>
+                        </span>
+                      </div>
+                    </div>
+                  </div>
+                {% else %}
+                  <div class="logged-out">
+                    <h3>{% trans "Let's get your account setup." %}</h3>
+                    <a class="button button--primary" href="{% url "github:github_auth" %}?redirect_uri={{ request.get_full_path }}"
```
Comment (mbeacom) [inline app/dashboard/templates/onboard.html]: This `url` view reference will need updated with the new views specified in `nav_auth.html`

File: app/dashboard/views.py [hunk 5]
```diff
@@ -426,6 +426,13 @@ def process_bounty(request):
 
     return TemplateResponse(request, 'process_bounty.html', params)
 
+def onboard(request):
+    "FTUX flow"
```
Comment (mbeacom) [inline app/dashboard/views.py]: You'll want to wrap this in triple double quotes. Something like: `"""Handle FTUX entry for onboard flow."""`

File: app/dashboard/templates/onboard.html [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+            <h2>{% trans "What's your skill set ?" %}</h2>
+            <div class="row">
+              <div class="col-12 offset-md-2 col-md-8 font-body pl-0 pr-0">
+                <p class="font-caption label">Suggested</p>
+                <div id="suggested-tags"></div>
+              </div>
+              <div class="search col-12 offset-md-2 col-md-8 pl-0 pr-0">
+                <p class="font-caption label">Search Skill</p>
+                <div class="search-area col-12 pl-0 pr-0">
+                  <input class="focus-effect form__input form__input_lg" id="keywords" type="text" placeholder="List out : Skill 1, Skill 2 ..." />
```
Comment (mbeacom) [inline app/dashboard/templates/onboard.html]: We might want to add `trans` to the placeholder text.

File: app/assets/v2/js/pages/onboard.js [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+var getFilters = function() {
+  var _filters = [];
+  var _words = [];
+  var search_keywords = $('#keywords').val();
+
+  if (search_keywords && search_keywords != '') {
+    search_keywords.split(',').forEach(function(word) {
+      _words.push(word);
+      _filters.push('<a class=filter-tag>' + word +
+    '<i class="fas fa-times" onclick="removeFilter(\'' + word + '\')"></i></a>');
```
Comment (mbeacom) [inline app/assets/v2/js/pages/onboard.js]: Can we indent this to match below?

File: app/assets/v2/js/pages/onboard.js [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+var removeFilter = function(value) {
+  $('input[name=tech-stack][value=' + value + ']').prop('checked', false);
+  getFilters();
+};
+
+var redirectURL = function() {
+  var level = $('#experienceLevel').find(':selected').text();
+  var url = '/explorer?q=' + words.join(',') + '&experience_level=' + level;
+  // TODO : Figure out how to set experience filter post re-direction to explorer
```
Comment (mbeacom) [inline app/assets/v2/js/pages/onboard.js]: 😄

File: app/assets/v2/images/metamask.svg (hunk 1 above)
Comment (thelostone-mc) [reply app/assets/v2/images/metamask.svg]: Will do ^_^

File: app/assets/v2/js/pages/onboard.js (hunk 7 above)
Comment (thelostone-mc) [reply app/assets/v2/js/pages/onboard.js]: Derp :P how did I miss that

Comment (mbeacom) [reply app/assets/v2/js/pages/onboard.js]: 😛

**Comment by thelostone-mc** [comment]
@mbeacom addressed all of them ^_^

File: app/assets/v2/js/pages/onboard.js [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+  $($('.step')[num]).addClass('block').outerWidth();
+  $($('.step')[num]).addClass('show');
+
+  if (num == 0)
+    $('#prev-btn').hide();
+  else
+    $('#prev-btn').show();
+
+  if (num == ($('.step').length) - 1) {
+    $('#next-btn').html('Finish');
```
Comment (kziemianek) [inline app/assets/v2/js/pages/onboard.js]: is this going to be translated somehow?

File: app/assets/v2/js/pages/onboard.js [hunk 10]
```diff
@@ -0,0 +1,164 @@
+var current = 0;
+var words = [];
+
+$('.js-select2').each(function() {
+  $(this).select2();
+});
+
+var showTab = function(num) {
```
Comment (kziemianek) [inline app/assets/v2/js/pages/onboard.js]: seems like global namespace pollution, maybe wrap in IIFE ?

File: app/assets/v2/js/pages/onboard.js (hunk 9 above)
Comment (thelostone-mc) [reply app/assets/v2/js/pages/onboard.js]: Ah don't think it will! @mbeacom ?

File: app/assets/v2/js/pages/onboard.js (hunk 10 above)
Comment (thelostone-mc) [reply app/assets/v2/js/pages/onboard.js]: A named IIFE  (cause I'm invoking it in the file later again )? 
Question : 
Wouln't a named IIFE end up polluting the global namespace too ? (I've used it only when I've wanted immediate invocation.)

**Comment by thelostone-mc** [comment]
@PixelantDesign Right now after you select the skill set -> I just redirect them to the explorer
We could add an alert there which shows them the instructions on the explorer itself

Would it make sense to add a Step4 where we show them the issues on this onboarding page rather than redirecting them to the explorer ? We'd have to re-use  bit of the code from the explorer to display the issues 
@mbeacom thoughts ?

File: app/assets/v2/js/pages/onboard.js (hunk 10 above)
Comment (thelostone-mc) [reply app/assets/v2/js/pages/onboard.js]: Oh thank for the lesson :D
makes sense on the long run 
@kziemianek 
Does this look ok ? 
https://gist.github.com/thelostone-mc/51d9d649d8802a823051c45a4ed886b8

**Comment by thelostone-mc** [comment]
@mbeacom @PixelantDesign this is ready for review ^_^

File: app/assets/v2/js/pages/onboard.js (hunk 9 above)
Comment (mbeacom) [reply app/assets/v2/js/pages/onboard.js]: As it sits currently, it won't. Should we wrap this in `gettext`?

File: app/assets/v2/js/pages/dashboard.js [hunk 11]
```diff
@@ -411,8 +411,54 @@ var getNextDayOfWeek = function(date, dayOfWeek) {
...
+      if (tag[j].value == 'any')
+        $('input[name=' + key + '][value=any]').prop('checked', true);
+      else
+        $('input[name=' + key + '][value=' + tag[j].value + ']').prop('checked', false);
+    }
+  }
+};
+
+(function() {
+  // if (document.referrer != 'http://0.0.0.0:8000/onboard') {
```
Comment (mbeacom) [inline app/assets/v2/js/pages/dashboard.js]: Can we drop this?

File: app/dashboard/templates/dashboard.html [hunk 12]
```diff
@@ -32,24 +32,42 @@
...
+                </div>
+                <div class="col-sm-5 col-md-7 bounty-info" id="funding-info">
+                </div>
               </div>
             </div>
           </div>
+          <div id="onboard-dashboard" class="row hidden font-body">
+            <div class="col-12 offset-md-1 col-md-9">
+              <h2 class="font-title">Welcome to Gitcoin!</h2>
+              <h3 class="font-subheader">We've found bounties that match your interest and experience</h3>
```
Comment (mbeacom) [inline app/dashboard/templates/dashboard.html]: We'll want to wrap all of the copy throughout this file with `{% trans %}` and `{% blocktrans %}` where applicable.

File: app/dashboard/views.py [hunk 13]
```diff
@@ -446,6 +446,13 @@ def process_bounty(request):
 
     return TemplateResponse(request, 'process_bounty.html', params)
 
+def onboard(request):
```
[the remark this answers is not in the document]
Comment (mbeacom) [reply app/dashboard/views.py]: You'll want another new line between `     return TemplateResponse(request, 'process_bounty.html', params)` and `def onboard(request):`

File: app/app/urls.py [hunk 14]
```diff
@@ -64,6 +64,7 @@
     url(r'^universe/(?P<issuenum>.*)/(?P<slug>.*)/?', external_bounties.views.external_bounties_show, name='universe'),
     url(r'^universe/?', external_bounties.views.external_bounties_index, name="universe_index"),
 
+    url(r'^onboard/?', dashboard.views.onboard, name='onboard'),
```
Comment (mbeacom) [inline app/app/urls.py]: Moving forward, we'll want to use `path` and `re_path` in place of `url`.

File: app/assets/v2/js/pages/dashboard.js (hunk 11 above)
Comment (thelostone-mc) [reply app/assets/v2/js/pages/dashboard.js]: I meant to drop that :P

**Comment by thelostone-mc** [comment]
@mbeacom all changes requested have been addressed in a neat lil PR ^_^

File: app/dashboard/templates/dashboard.html [hunk 15]
```diff
@@ -32,24 +32,42 @@
...
+                  <span id="matches" class="font-subheading"></span>
+                </div>
+                <div class="col-sm-5 col-md-7 bounty-info" id="funding-info">
+                </div>
               </div>
             </div>
           </div>
+          <div id="onboard-dashboard" class="row hidden font-body">
+            <div class="col-12 offset-md-1 col-md-9">
+              <h2 class="font-title">{% trans Welcome to Gitcoin! %}</h2>
```
Comment (mbeacom) [inline app/dashboard/templates/dashboard.html]: You'll want to wrap the text being translated in `"`

**Comment by mbeacom** [comment]
One minor change request regarding translation support.  Other than that, lgtm.
Thanks @thelostone-mc !

**Comment by owocki** [comment]
how does the user get to /onboard page?
 - we need a link on the landing page (maybe it can replace the 'tools' page?)
 - we also need a way for existing users who haven't already filled out their profile to go through onboarding



entire flow -- the URL doesnt seem to be changing when i go from page to page.  what if i want to go back to a specific spot on the flow?

page 1
looks kind of funny in tablet view http://bits.owocki.com/1u3B1A2C0I2A/Screen%20Shot%202018-04-23%20at%205.54.27%20PM.png
and in mobile view it could maybe use a HR between the build/paid/network thing and the 'youve logged in' thing http://bits.owocki.com/1V0n0i0o3F1S/Screen%20Shot%202018-04-23%20at%205.54.46%20PM.png

page 2
tablet view is kind of screwed up http://bits.owocki.com/3U1o2P1q1p2h/Screen%20Shot%202018-04-23%20at%205.55.42%20PM.png
shouldnt the 'prev' and 'next' links be buttons like in the design? 

when i go to the metamask URL and then click back it lands me back on page 1.  it should take me to  the metamask page (or better yet, i should never leave)

after i install metamask, the 'unlock metamask' button appears but it still takes me to the metamask homepage when i click it.  maybe it should alert me how to unlock metamask intaed?


none of these links work
http://bits.owocki.com/0x171l3T1T3E/Screen%20Shot%202018-04-23%20at%205.59.56%20PM.png

page 3

skills dont seem to save

also, they take me to this 500 page when i click finish
http://bits.owocki.com/3S142Y2B2u0d/Screen%20Shot%202018-04-23%20at%206.00.16%20PM.png



@thelostone-mc my qa notes above ^^

**Comment by PixelantDesign** [comment]
> how does the user get to /onboard page?
> 
> we need a link on the landing page (maybe it can replace the 'tools' page?)
> we also need a way for existing users who haven't already filled out their profile to go through onboarding

*New User* will arrive via link from the landing page (TBDesigned).
*Existing User* will get a prompt when the next time they come back to the issue explorer. We won't be able to tell who is a funder or who is a contributor (unless there is a way we can hand select the funders that we so that they get the funder ftux. Thoughts?

@owocki @thelostone-mc mentioned to me earlier today that nothing is wired up (just yet).

**Comment by thelostone-mc** [comment]
@owocki I sort of changed the design cause I had some free time just to see if I could update it 

Spoke to Alisa offline, will get it to look like the original flow 

I was waiting on feedback for the updates design , figured I'd optimize for mobile + tablet after that 

The explorer should work , I'll double check that

**Comment by thelostone-mc** [comment]
Ideally for onboarding usecase => everything resides on one page , do we want to allow to jump to a specific step at any point of time ? ( If so we can split it across as multiple links but each time the user clicks next -> the whole page refreshes as opposed to the current setup where it's a subtle change ) 

Thoughts ?

**Comment by thelostone-mc** [comment]
@owocki It should work now 
I ended up using the translate incorrectly 😅 
Expected: `{% trans "GOT IT !" %}`
What I wrote: `{% trans GOT IT ! %}`

I know @mbeacom is judging me for this , I have no excuses 😂

**Comment by owocki** [comment]
> Ideally for onboarding usecase => everything resides on one page , do we want to allow to jump to a specific step at any point of time ? ( If so we can split it across as multiple links but each time the user clicks next -> the whole page refreshes as opposed to the current setup where it's a subtle change )

Using HTML5 you can mutate the user's URL without making them leave the page.

**Comment by thelostone-mc** [comment]
@owocki will get that done! Thanks :D

**Comment by owocki** [comment]
very excited for this

qa notes:
* buttons in the design are positioned on left / right of page,  both are on right in implementation
* buttons are dark (sometimes) in design.  in implementation they are white
* if i've already gone through onboarding, shouldn't my skills / experience level be prefilled?
* it'd be nice to have autocomplete on the 'add skill' input bar like there is on the search bar on gitcoin.co/explorer
* build your avatar is stubbed in.. yay! 
* [positioning is all screwed up on this page](http://bits.owocki.com/0V0x3R3D3j02/screencapture-localhost-8000-explorer-2018-04-27-15_51_52.png). screen is like 950px wide
* it does not seem like the keywords i input during onboarding are actually saved in my profile because i dont see them when i go to http://localhost:8000/settings/matching
* speaking of the settings/matching page... the '(pull from github)' option could be cool on the onboard page

dont know how much of this is scope creep and what isnt... but let me know what you think when you view it @thelostone-mc !

**Comment by PixelantDesign** [comment]
@owocki @thelostone-mc

it does not seem like the keywords i input during onboarding are actually saved in my profile because i dont see them when i go to http://localhost:8000/settings/matching

Hmm did not know that had to be done ! Mind pointing me in the right direction ? 😅

*The above was not part of the scope. Lets make that v2, since we dont currently have a sesign of where to put that in the profile.*

**Comment by PixelantDesign** [comment]
@owocki Great catches on the details like buttons and things. We do have another task for visual design clean up when we stutch together the ftux and avatar builder.

**Comment by owocki** [comment]
> We do have another task for visual design clean up when we stutch together the ftux and avatar builder.

okie dokie

> The above was not part of the scope. Lets make that v2, since we dont currently have a sesign of where to put that in the profile.

i'm fine with that as long as it gets done before we launch.  this is important data for us, and i dont want to lose it

> Figured we'd add it once we get the taggit keywords set up ! (Thoughts @mbeacom ? )

maybe we can move it to v2?  idk, itd be easy to copy from the search bar implementation quite frankly.

> Where would we get this info from? We can auto-populate it

EmailSubscriber.keywords

> Could you tell me which one?

designs on https://github.com/gitcoinco/web/issues/441 - i dont have a detailed button inventory, but i was talking about the 'next' and 'previous' buttons in the bottom left / right

**Comment by owocki** [comment]
@thelostone-mc @PixelantDesign is there a v2 ticket somewhere? id love to get that scoped up and bountied soon so we can keep this rollin'

@mbeacom do we want to hide the `/onboard` page in prod until its done?

**Comment by thelostone-mc** [comment]
@owocki focussing on integrating the avatar to the onboarding flow with @Smona here https://github.com/gitcoinco/web/pull/967

I believe @PixelantDesign has a ticket open for v2 (not sure which one)
Parallely will work on saving the skill set for the new user & auto populate in for exisiting users

**Pull request merged**